utils/swift_build_support/swift_build_support/build_graph.py

```diff
@@ -0,0 +1,125 @@
+# swift_build_support/build_graph.py ----------------------------*- python -*-
+#
+# This source file is part of the Swift.org open source project
+#
+# Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+# Licensed under Apache License v2.0 with Runtime Library Exception
+#
+# See https://swift.org/LICENSE.txt for license information
+# See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+#
+# ----------------------------------------------------------------------------
+#
+# This is a simple implementation of an acyclic build graph. We require no
+# cycles, so we just perform a reverse post order traversal to get a topological
+# ordering. We check during the reverse post order traversal that we do not
+# visit any node multiple times.
+#
+# Nodes are assumed to be a product's class.
+#
+# ----------------------------------------------------------------------------
+
+class BuildDAG(object):
+
+    def __init__(self):
+        self.root = None
+
+        # A map from a node to a list of nodes that the node depends on.
+        #
+        # As such, one can never add a dependence of anything on root, so we
+        # assert if one attempts to due that.
+        self.invertedDepMap = {}
+
+    def add_edge(self, pred, succ):
+        self.invertedDepMap.setdefault(pred, set([succ])) \
+                           .update([succ])
+
+    def set_root(self, root):
+        assert(self.root is None)
+        self.root = root
+
+    def produce_schedule(self):
+        # Ok, we have the root of the graph. Lets construct a post order
+        # map from keys -> numbers.
+        root = self.root
+        worklist = [root]
+        visitedNodes = set([])
+
+        po_ordered_nodes = []
+        while not len(worklist) == 0:
+            node = worklist[-1]
+            if node in visitedNodes:
+                worklist.pop()
+                continue
+
+            deps = self.invertedDepMap.get(node, set([]))
+            assert(isinstance(deps, set))
+
+            foundDep = False
+            for d in deps:
+                if d not in visitedNodes:
+                    foundDep = True
+                    worklist.append(d)
+            if foundDep:
+                continue
```

Comment:
I'm not sure I understand to purpose of the `continue` here. Wouldn't we just end up going beyond this in the next step of the iteration? In case I am missing something here, I think a comment would be useful.

Reply:
I want to visit the children before I visit myself. I'll add a comment.

Reply:
Oh, I see. Just realized that `list[-1]` returns the last element. This absolutely makes sense, thanks.